File: mfga-xposed/app/src/main/java/com/mfga/xposed/ReplacementGuard.java
package com.mfga.xposed;

import java.util.function.Supplier;

/** Per-thread recursion guard; original nulls and exceptions are not turned into success. */
public final class ReplacementGuard {
    private static final ThreadLocal<Boolean> ACTIVE = new ThreadLocal<>();

    private ReplacementGuard() {}

    public static boolean isActive() {
        return Boolean.TRUE.equals(ACTIVE.get());
    }

    public static <T> T replace(T original, Supplier<T> factory) {
        if (original == null || isActive()) return original;
        ACTIVE.set(Boolean.TRUE);
        try {
            T replacement = factory.get();
            return replacement != null ? replacement : original;
        } finally {
            ACTIVE.remove();
        }
    }
}

File: mfga-xposed/app/src/main/kotlin/com/mfga/xposed/ReplacementGuard.kt
package com.mfga.xposed

import java.util.function.Supplier

/** Per-thread recursion guard; original nulls and exceptions are not turned into success. */
object ReplacementGuard {
    private val ACTIVE = ThreadLocal<Boolean>()

    @JvmStatic
    fun isActive(): Boolean = ACTIVE.get() == true

    @JvmStatic
    fun <T> replace(original: T?, factory: Supplier<T>): T? {
        if (original == null || isActive()) return original
        ACTIVE.set(true)
        return try {
            val replacement = factory.get()
            replacement ?: original
        } finally {
            ACTIVE.remove()
        }
    }
}
